src/test/fio_ceph_objectstore.cc

```diff
@@ -0,0 +1,345 @@
+
+	return ceph_os_data->aio_events[event];
+}
+
+static int fio_ceph_os_getevents(struct thread_data *td, unsigned int min,
+			     unsigned int max, const struct timespec *t)
+{
+	struct ceph_os_data *ceph_os_data = (struct ceph_os_data *) td->io_ops->data;
+	unsigned int events = 0;
+	struct io_u *io_u;
+	unsigned int i;
+	struct fio_ceph_os_iou *fov;
+
+	do {
+		io_u_qiter(&td->io_u_all, io_u, i) {
+			if (!(io_u->flags & IO_U_F_FLIGHT))
+				continue;
+
+			fov = (struct fio_ceph_os_iou *)io_u->engine_data;
+
+			if (fov->io_complete) {
+				fov->io_complete = 0;
+				ceph_os_data->aio_events[events] = io_u;
+				events++;
+			}
+
+		}
+		if (events < min)
+			usleep(100);
+		else
+			break;
+
+	} while (1);
+
+	return events;
+}
+
+static int fio_ceph_os_queue(struct thread_data *td, struct io_u *io_u)
+{
+	int r = -1;
+	char buf[32];
+	struct ceph_os_data *ceph_os_data = (struct ceph_os_data *) td->io_ops->data;
+	uint64_t len = io_u->xfer_buflen;
+	uint64_t off = io_u->offset;
+	ObjectStore *fs = ceph_os_data->fs;
+	snprintf(buf, sizeof(buf), "XXX_%lu_%lu", io_u->start_time.tv_usec, io_u->start_time.tv_sec);
+	spg_t pg;
+	ghobject_t oid(pg.make_temp_object(buf));
+
+	fio_ro_check(td, io_u);
+
+	bufferlist data;
+	data.push_back(buffer::create_static(len, (char *)io_u->xfer_buf));
+
+        if (io_u->ddir == DDIR_WRITE) {
+		ObjectStore::Transaction *t = new ObjectStore::Transaction;
+		if (!t) {
+			cout << "ObjectStore Transcation allocation failed." << std::endl;
+			goto failed;
+		}
+		t->write(coll_t(pg), oid, off, len, data);
+		//cout << "QUEUING transaction " << io_u << std::endl;
+		fs->queue_transaction(NULL, t, new OnApplied(io_u, t), new OnCommitted(io_u));
+	} else if (io_u->ddir == DDIR_READ) {
+		r = fs->read(coll_t(pg), oid, off, len, data);
+		if (r < 0)
+			goto failed;
+		io_u->resid = len - r;
+		return FIO_Q_COMPLETED;
+	} else {
+		cout << "WARNING: No DDIR beside DDIR_WRITE supported!" << std::endl;
+		return FIO_Q_COMPLETED;
+	}
+
+	return FIO_Q_QUEUED;
+
+failed:
+	io_u->error = r;
+	td_verror(td, io_u->error, "xfer");
+	return FIO_Q_COMPLETED;
+}
+
+static int fio_ceph_os_init(struct thread_data *td)
+{
+	vector<const char*> args;
+	struct ceph_os_data *ceph_os_data = (struct ceph_os_data *) td->io_ops->data;
+	struct ceph_os_options *o = (struct ceph_os_options *) td->eo;
+	ObjectStore::Transaction ft;
+
+	global_init(NULL, args, CEPH_ENTITY_TYPE_OSD, CODE_ENVIRONMENT_UTILITY, 0);
+	//g_conf->journal_dio = false;
+	common_init_finish(g_ceph_context);
+	if (o->filestore_debug) {
+		g_ceph_context->_conf->set_val("debug_filestore", o->filestore_debug);
+		g_ceph_context->_conf->apply_changes(NULL);
+	}
+
+  ObjectStore *fs = ObjectStore::create(g_ceph_context,
+			o->objectstore, td->o.directory,
+			o->filestore_journal ? o->filestore_journal : "");
+	if (fs == NULL) {
+		cout << "bad objectstore type " << o->objectstore << std::endl;
+		return 1;
+	}
+	if (fs->mkfs() < 0) {
```

Comment:
I implemented fio's close_file() hook by removing the associated objects, so it should clean up after itself now.